### PR TITLE
Implements extraction of Archives before Scanning

### PR DESCRIPTION
**Pull request by ashutoshsaboo**
Signed-off-by: Ashutosh Saboo <ashutosh.saboo96@gmail.com>

@pombredanne Sir. This is a step towards implementing extraction of archives for scanning of licenses within the files of the archives as well. 

I feel, instead of extracting to a temporary location, as @pombredanne you were discussing in #14 , we must instead extract to the original location itself, since that will also not hinder the performance of the scanning engine as well. Scancode then itself, searches for licenses in the extracted files, and then we could delete the extracted files as well.

I have been looking at Scancode's source code since 2-3 days now, and I would be willing to work for implementing this. Any suggestions on this @pombredanne Sir? I'll issue new commits with the remaining TODOs as well.

Achieved:
1. Extraction of all archives in the folder tree under a specified location before starting the scanning process.
2. Silent extraction of archives.

Currently TODO's:
1. Delete extracted files after scanning is complete
2. Also implement points mentioned in #458 for specific archive extraction.

### CODE REVIEW

**Comment by pombredanne** [comment]
Thanks for the input. This should go to a temporary location and not in-place+delete.
Also it should be implemented as an alternative resources iterator that receives a stream of resources and handles extraction on the fly if needed and yields more resources. So it should be its own function. I suggest you start with tests too.

**Comment by ashutoshsaboo** [comment]
@pombredanne I am sorry, could you explain what do you mean by temporary location - not in - place+delete? I understand your alternate resources iterator point though. I'll try working towards it.

**Comment by ashutoshsaboo** [comment]
I was thinking of something of this kind, for a revamped iterator that'll handle, extracting the files part too. 

1)But then when we try to `import extractcode` within the `fileutils.py` file, it pops with an error. Any suggestion on how to combat that? 

2)And just one thing, the extract() function actually gives returns an iterable. Is there any functionality implemented as of now, that'll directly extract the location that we specify recursively, without returning the iterable, because then we can avoid the unnecessary printing of extract statements on the terminal ?


```
def resource_iter_extractable(location, ignored=ignore_nothing, with_files=True, with_dirs=True, with_extract=True):
    """
    Return an iterable of resources at `location` recursively.

    :param location: a file or a directory.
    :param ignored: a callable accepting a location argument and returning True
                    if the location should be ignored.
    :param with_dirs: If True, include the directories.
    :param with_files: If True, include the  files.
    :return: an iterable of file and directory locations.
    """
    assert  with_dirs or with_files, "One or both of 'with_dirs' and 'with_files' is required"
    for top, dirs, files in walk(location, ignored):
        if with_files:
            for f in files:
                if archive.can_extract(f) and with_extract:
                    for y in extract(f, kinds=default_kinds, recurse=True):
                        print (y)
                    yield os.path.join(top,f+'-extract')
                else:
                    yield os.path.join(top, f)
        if with_dirs:
            for d in dirs:
                yield os.path.join(top, d)

```

**Comment by pombredanne** [comment]
do not modify fileutils. You are creating circular imports. extracting and print are unrelated. Check the code in extractcode

**Comment by ashutoshsaboo** [comment]
@pombredanne Yes I already checked the code in extractcode. I'm looking into it. One thing, I noticed after I ran the test cases on my local system is that, the reason for the failing tests was basically related to the archives. It basically compares the output of Scancode, with a pre-defined json usually saved in the info folder. Now it contains, params like number of files scanned and others as well, which are bound to increase after we, scan the files in the archives as well. 

So, I'll have to change the test cases first, by modifying the pre-defined jsons. So, that seems to be correct right? @pombredanne

**Comment by pombredanne** [comment]
You did not get my point:  start a new script in `src/extractcode` ... do not modify existing code for now. Create new tests as needed.

**Comment by ashutoshsaboo** [comment]
@pombredanne Thanks for your input. I've added a new script that achieves extracting all the archives within a directory, before starting the scanning process. Please provide your input on the same. 

I've also updated the TODOs in the main issue, which I'll be working on shortly.

**Comment by ashutoshsaboo** [comment]
I just checked out your wiki here - https://github.com/nexB/aboutcode/wiki/GSOC-2017 . I'd like to work on this full time, to resolve #14 completely, if that's fine. Would that be fine @pombredanne ?

File: src/extractcode/extract_directory.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+#
+# Copyright (c) 2015 nexB Inc. and others. All rights reserved.
```
Comment (pombredanne) [inline src/extractcode/extract_directory.py]: I guess we are no longer on 2015, are we?
Be careful to not copy extra things around without checking first.

File: src/extractcode/extract_directory.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+TRACE = False
+
+if TRACE:
+    import sys
+    logging.basicConfig(stream=sys.stdout)
+    logger.setLevel(logging.DEBUG)
+
+
+"""
+Extract archives and compressed files recursively to get the file content available for
```
Comment (pombredanne) [inline src/extractcode/extract_directory.py]: How is this doc relevant to this new script? Pleas remove

File: src/scancode/cli.py [hunk 3]
```diff
@@ -70,6 +70,11 @@
 
 from scancode import utils
 
+from extractcode.extract import extract
+from extractcode import default_kinds
+from extractcode import archive
+from extractcode.extract_directory import extract_directory
+
```
Comment (pombredanne) [inline src/scancode/cli.py]: None of these import are useful at this stage

File: src/extractcode/extract_directory.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    Note that while the original file system is walked top-down, breadth-first,
+    if recurse and a nested archive is found, it is extracted to full depth
+    first before resuming the file system walk.
+    """
+    abs_input = fileutils.as_posixpath(os.path.abspath(os.path.expanduser(location)))
+    original_input = fileutils.as_posixpath(location)
+    abs_input = fileutils.as_posixpath(os.path.abspath(os.path.expanduser(location)))
+    ignored = partial(ignore.is_ignored, ignores=ignore.ignores_VCS, unignores={})
+    resources = fileutils.resource_iter(abs_input, ignored=ignored)
```
Comment (pombredanne) [inline src/extractcode/extract_directory.py]: I am confused as to why you are creating a resource iterator here. This does not sound right.
A transparent extraction iterator would receive a stream or resources and yield more resources after extraction.

File: src/scancode/cli.py (hunk 3 above)
Comment (ashutoshsaboo) [inline src/scancode/cli.py]: Resolved this, in the latest commit. :+1:

File: src/extractcode/extract_directory.py (hunk 2 above)
Comment (ashutoshsaboo) [inline src/extractcode/extract_directory.py]: Resolved this, in the latest commit. :+1:

File: src/extractcode/extract_directory.py (hunk 1 above)
Comment (ashutoshsaboo) [inline src/extractcode/extract_directory.py]: Oops. Sorry for the mistake. Resolved this, in the latest commit. :+1:

File: src/extractcode/extract_directory.py (hunk 4 above)
Comment (ashutoshsaboo) [inline src/extractcode/extract_directory.py]: Resolved this in the latest commit as well. :+1: Thanks for your input. :smile:

**Comment by ashutoshsaboo** [comment]
@pombredanne thanks a lot for your latest comment here #14 . I have pushed a few more commits to this PR , which works on basically accepting a resources iterator as input, and extract all the files that are extractible before beginning with the scanning process. This works in the short run as of now, and it suffices the need very well, because the latest commit, will basically take the resource iterator as an input, and extract all the extractible files, to ensure that the tree is built perfectly during scanning.

I understand your point, about the File object thing as well. I have been thinking of the same thing for my GSoC proposal as well, since implementing that with the correct unit tests should take a considerable amount of time as well. 

Basically once, we extract an archive we'll have to report both the **real path** as well as the **extracted/virtual path** as well. So, I'll think about implementing that functionality too before writing my proposal. 

Any other inputs from your end, which I could include in my proposal? I'll communicate the points that I have thought of as well, shortly. Also, is there any other requirement that you would want me to meet before sending the proposal? I have been in good touch with the code-base in general, and I've pushed #544 and #547 till now. I'll be working pushing more commits shortly as well.

Is there any specific/additional thing that you would want me to work towards to be a worthy GSoC candidate for resolving #14 completely? I am open to working full time for this as well. 

Would be good to have your input on this! :smile:

**Comment by pombredanne** [comment]
The approach you detailed in https://github.com/nexB/scancode-toolkit/pull/544#issuecomment-284671722 seems fine. You just need to not print things. instead you further recursively iterate once you have extracted something.

**Comment by pombredanne** [comment]
The outline of what do to would be:
```
def iter_extract(resources)
    for resource in resources:
        if is extractible:
            extract(target  dir)
            return iter_extract(target dir)
        else:
            yield resource
```

**Comment by pombredanne** [comment]
> Is there any specific/additional thing that you would want me to work towards to be a worthy GSoC candidate for resolving #14 completely? I am open to working full time for this as well.

You are almost there. So finishing this would make sense if you can.

**Comment by pombredanne** [comment]
ping?

**Comment by ashutoshsaboo** [comment]
@pombredanne Caught up with exams. Will work on this ASAP, the exams get over. Thanks :+1: .

**Comment by ashutoshsaboo** [comment]
@pombredanne Sorry for the delay, I'm caught up with exams, which is taking most of my time. I've tried to make some changes, to make the new extract iterator yield extracted files also on the fly. Please let me know your feedback on the latest commit. Thanks :smile:

**Comment by ashutoshsaboo** [comment]
@pombredanne Why do the Appveyor tests fail, while the Travis tests pass? Could you help with this?
Also, any comments on the recent commit? Seems fine?

**Comment by pombredanne** [comment]
The tests failure are not yours but mine!

File: src/scancode/cli.py [hunk 5]
```diff
@@ -280,6 +285,24 @@ def scancode(ctx,
     scans_cache_class = get_scans_cache_class()
 
     try:
+        #We could call the below function if required to first extract all the archives, and then start scanning
+
+        '''
+        abs_input = fileutils.as_posixpath(os.path.abspath(os.path.expanduser(location)))
```
Comment (pombredanne) [inline src/scancode/cli.py]: What is this comment code?

Comment (ashutoshsaboo) [inline src/scancode/cli.py]: Yes I'll remove it.

Comment (ashutoshsaboo) [inline src/scancode/cli.py]: Removed in the latest commit. :+1:

**Comment by ashutoshsaboo** [comment]
@pombredanne Let me know once your tests pass, and if this looks perfect? :smile:

File: src/scancode/cli.py [hunk 6]
```diff
@@ -280,6 +285,22 @@ def scancode(ctx,
     scans_cache_class = get_scans_cache_class()
 
     try:
+        '''
```
Comment (pombredanne) [inline src/scancode/cli.py]: Why put a string instead of code? Please either delete or implement it

**Comment by ashutoshsaboo** [comment]
Should I add a separate option called ``--extract`` that would search for licenses within the archives as well? Since, in case if we don't add any such option, it'd require modifying all the other tests completely, which can lead to failing tests as well, and cause other problems. So, adding a new option like ``--extract`` would seem to be fine? @pombredanne

**Comment by pombredanne** [comment]
@ashutoshsaboo a new option is fine. You should model it after the extractcode cli options IMHO

**Comment by ashutoshsaboo** [comment]
@pombredanne Thanks a lot. I'll work on it some time shortly. :smile:

**Comment by pombredanne** [comment]
@ashutoshsaboo I am leaving this open as I might cherry pick some of it later when we come to implement this for real (eventually as a plugin).

**Comment by majurg** [comment]
@ashutoshsaboo @pombredanne Ping. Not sure if we want this open still or if I should close.

**Comment by pombredanne** [comment]
I am closing this at this stage, but I kept a copy of your branch here in https://github.com/nexB/scancode-toolkit/tree/ashutoshsaboo/archive_extraction so we can reuse it when we come to implement this after all!.
Thank you ++
